Port to STM32

STM32 are powerful 32 bits MCUs, some of them have enough resources to run AtomVM.
It would be nice to have support for them.

Some progress: the build now fails at link time (as it should, since I didn't implement the sys module yet) but in the process of doing so, it spits out a bunch of warnings:
```
  LD    atom-vm.elf
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: error: atom-vm.elf uses VFP register arguments, ../libatom-vm/build/liblibAtomVM.a(globalcontext.c.obj) does not
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: failed to merge target specific data of file ../libatom-vm/build/liblibAtomVM.a(globalcontext.c.obj)
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: error: atom-vm.elf uses VFP register arguments, ../libatom-vm/build/liblibAtomVM.a(valueshashtable.c.obj) does not
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: failed to merge target specific data of file ../libatom-vm/build/liblibAtomVM.a(valueshashtable.c.obj)
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: error: atom-vm.elf uses VFP register arguments, ../libatom-vm/build/liblibAtomVM.a(atom.c.obj) does not
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: failed to merge target specific data of file ../libatom-vm/build/liblibAtomVM.a(atom.c.obj)
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: error: atom-vm.elf uses VFP register arguments, ../libatom-vm/build/liblibAtomVM.a(atomshashtable.c.obj) does not
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: failed to merge target specific data of file ../libatom-vm/build/liblibAtomVM.a(atomshashtable.c.obj)
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-abort.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-calloc.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-fprintf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-impure.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-malloc.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memcmp.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memcpy.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-callocr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-freer.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-mallocr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-vfprintf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-vfprintf_i.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-puts.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-sbrkr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-signal.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-signalr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-strlen.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-strnlen.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-wbuf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-wsetup.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-fflush.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-findfp.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-fvwrite.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-fwalk.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-makebuf.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memchr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memmove.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-memset.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-reallocr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-reent.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-stdio.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-writer.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-closer.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-fstatr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-isattyr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-lseekr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-nano-msizer.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
/usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/bin/ld: warning: /usr/lib/gcc/arm-none-eabi/4.9.3/../../../arm-none-eabi/lib/armv7e-m/fpu/libc_nano.a(lib_a-readr.o) uses 2-byte wchar_t yet the output is to use 4-byte wchar_t; use of wchar_t values across objects may fail
../libatom-vm/build/liblibAtomVM.a(globalcontext.c.obj): In function `globalcontext_get_module':
/home/rbino/PersonalProjects/AtomVM/src/libAtomVM/globalcontext.c:227: undefined reference to `sys_load_module'
collect2: error: ld returned 1 exit status
../rules.mk:135: recipe for target 'atom-vm.elf' failed
make: *** [atom-vm.elf] Error 1
```

--- a/examples/elixir/stm32/Blink.ex
+++ b/examples/elixir/stm32/Blink.ex
@@ -1,0 +1,14 @@
+defmodule Blink do
+  def start do
+    spawn(Blinker, :start, [{:d, 12}, 1000])
+    spawn(Blinker, :start, [{:d, 13}, 500])
+    spawn(Blinker, :start, [{:d, 14}, 1500])
+    spawn(Blinker, :start, [{:d, 15}, 300])
+
+    loop()
+  end
+
+  def loop do
+    loop()
+  end
+end

--- a/examples/elixir/stm32/Blinker.ex
+++ b/examples/elixir/stm32/Blinker.ex
@@ -1,0 +1,16 @@
+defmodule Blinker do
+  def start(gpio, interval_ms) do
+    gpio_driver = GPIO.open();
+    GPIO.set_direction(gpio_driver, gpio, :output)
+
+    loop(gpio_driver, gpio, interval_ms, 0)
+  end
+
+  def loop(gpio_driver, gpio, interval_ms, level) do
+    GPIO.set_level(gpio_driver, gpio, level)
+
+    :timer.sleep(interval_ms)
+
+    loop(gpio_driver, gpio, interval_ms, 1 - level)
+  end
+end

--- a/examples/elixir/stm32/README.md
+++ b/examples/elixir/stm32/README.md
@@ -1,0 +1,13 @@
+# STM32F4Discovery example
+
+This example will blink the 4 LEDs present on the STM32F4Discovery with
+different frequencies, using one process per LED.
+
+# Building
+- Run `./build.sh`
+- If you haven't done so, build PackBEAM (in the `tools` directory)
+- Use PackBEAM to pack all Beam module: `../../../tools/packbeam/build/PackBEAM blink-stm32.avm *.beam`
+
+# Flashing
+- Follow the STM32 README (`src/platforms/stm32`) to build and flash AtomVM
+- Flash the packed AVM with `st-flash --reset write blink-stm32.avm 0x8080000`

--- a/examples/elixir/stm32/build.sh
+++ b/examples/elixir/stm32/build.sh
@@ -1,0 +1,4 @@
+#!/bin/bash
+elixirc *.ex
+erlc ../../../libs/estdlib/*.erl
+elixirc ../../../libs/exavmlib/*.ex

--- a/src/platforms/stm32/.gitignore
+++ b/src/platforms/stm32/.gitignore
@@ -1,0 +1,4 @@
+build/
+*.elf
+*.bin
+generated.*.ld

--- a/src/platforms/stm32/CMakeLists.txt
+++ b/src/platforms/stm32/CMakeLists.txt
@@ -40,6 +40,10 @@ include_directories(src)
 # Specify output executable
 add_executable(${PROJECT_NAME}.elf ${SOURCE_FILES})
 
+add_subdirectory(../../libAtomVM ./libAtomVM)
+target_include_directories(${PROJECT_NAME}.elf PRIVATE ../../libAtomVM)
+target_link_libraries(${PROJECT_NAME}.elf "libAtomVM")
+
 # Output elf file size
 add_custom_command(
     TARGET ${PROJECT_NAME}.elf

--- a/src/platforms/stm32/CMakeLists.txt
+++ b/src/platforms/stm32/CMakeLists.txt
@@ -1,0 +1,49 @@
+cmake_minimum_required(VERSION 3.9)
+project(atom-vm)
+
+# Include an error in case the user forgets to specify ARM as a toolchain
+if (NOT CMAKE_TOOLCHAIN_FILE)
+    message(FATAL_ERROR "Cross compiling only. Please use -DCMAKE_TOOLCHAIN_FILE=cmake/arm-toolchain.cmake or use\
+    your own toolchain file")
+endif ()
+
+if ((NOT ${CMAKE_C_COMPILER_ID} STREQUAL "GNU") OR
+    (NOT ${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU") OR
+    (${CMAKE_CXX_COMPILER_VERSION} VERSION_LESS 7.2.1))
+    message(WARNING "This project is targeted for ARM GCC versions 7.2.1 or above. Some features may not be supported\
+     by your compiler. You are using:
+     ${CMAKE_CXX_COMPILER_ID} C, version ${CMAKE_C_COMPILER_VERSION}
+     ${CMAKE_CXX_COMPILER_ID} CXX, version ${CMAKE_CXX_COMPILER_VERSION}")
+endif()
+
+# Set the C/C++ standards
+set(CMAKE_C_STANDARD 11)
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/build)
+
+# Verbose logging
+set(LOG_VERBOSE false)
+
+# Choose target device here
+set(DEVICE stm32f407vgt6)
+
+# Include libopencm3
+include(cmake/libopencm3.cmake)
+
+# Include additional compilation flags
+include(cmake/compile-flags.cmake)
+
+# Collect sources into SOURCE_FILES variable and include your headers directory
+file(GLOB_RECURSE SOURCE_FILES "src/*.c" "src/*.cpp")
+include_directories(src)
+
+# Specify output executable
+add_executable(${PROJECT_NAME}.elf ${SOURCE_FILES})
+
+# Output elf file size
+add_custom_command(
+    TARGET ${PROJECT_NAME}.elf
+    POST_BUILD
+    COMMAND ${ARM_SIZE} ${PROJECT_NAME}.elf
+    WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}
+)

--- a/src/platforms/stm32/CMakeLists.txt
+++ b/src/platforms/stm32/CMakeLists.txt
@@ -24,8 +24,9 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/build)
 # Verbose logging
 set(LOG_VERBOSE false)
 
-# Choose target device here
-set(DEVICE stm32f407vgt6)
+if (NOT DEVICE)
+    set(DEVICE stm32f407vgt6)
+endif ()
 
 # Include libopencm3
 include(cmake/libopencm3.cmake)

--- a/src/platforms/stm32/README.md
+++ b/src/platforms/stm32/README.md
@@ -5,3 +5,7 @@
 - `cd build`
 - `cmake -DCMAKE_TOOLCHAIN_FILE=../cmake/arm-toolchain.cmake -DLIBOPENCM3_DIR=../libopencm3 ..`
 - `make`
+
+The default build is based on the STM32F4Discovery board chip. If you want to target a different chip,
+pass the `-DDEVICE` flag when invoking cmake. For example, to use the STM32F429Discovery,
+pass `-DDEVICE=stm32f429zit6`

--- a/src/platforms/stm32/README.md
+++ b/src/platforms/stm32/README.md
@@ -1,0 +1,7 @@
+# AtomVM for STM32
+- `git clone libopencm3`
+- `cd libopencm3 && make -j4 && cd ..`
+- `mkdir build`
+- `cd build`
+- `cmake -DCMAKE_TOOLCHAIN_FILE=../cmake/arm-toolchain.cmake -DLIBOPENCM3_DIR=../libopencm3 ..`
+- `make`

--- a/src/platforms/stm32/README.md
+++ b/src/platforms/stm32/README.md
@@ -1,4 +1,10 @@
 # AtomVM for STM32
+
+## Prerequisites
+To flash both AtomVM and your packed AVM you must use st-flash (https://github.com/texane/stlink). Make sure to
+follow its [installation procedure](https://github.com/texane/stlink#installation) before proceeding further.
+
+## Building
 - `git clone libopencm3`
 - `cd libopencm3 && make -j4 && cd ..`
 - `mkdir build`
@@ -6,6 +12,30 @@
 - `cmake -DCMAKE_TOOLCHAIN_FILE=../cmake/arm-toolchain.cmake -DLIBOPENCM3_DIR=../libopencm3 ..`
 - `make`
 
-The default build is based on the STM32F4Discovery board chip. If you want to target a different chip,
-pass the `-DDEVICE` flag when invoking cmake. For example, to use the STM32F429Discovery,
-pass `-DDEVICE=stm32f429zit6`
+You can put libopencm3 wherever you want on your PC as long as you update LIBOPENCM3_DIR to point to it.
+
+## Flashing
+To flash AtomVM, use
+```
+st-flash --reset write atom-vm.bin 0x8000000
+```
+
+To flash your packed AVM, use
+```
+st-flash --reset write /path/to/your/packed.avm 0x8080000
+```
+
+Right now AtomVM expects to find the AVM at the address 0x808000. On a STM32 Discovery board this means that
+the 1MB of flash will be split in 512KB available for the program and 512KB available for the packed AVM.
+If for any reason you want to modify this, you can change `AVM_ADDRESS` and `AVM_FLASH_MAX_SIZE` defines in `main.c`.
+
+# Printing
+By default, stdout and stderr are printed on USART2. On the STM32F4Discovery board, you can see them
+using a TTL-USB with the TX pin connected to board's pin PA2 (USART2 RX). Baudrate is 115200 and serial transmission
+is 8N1 with no flow control.
+
+## Changing device
+
+The default build is based on the STM32F4Discovery board chip (`stm32f407vgt6`). If you want to target a different
+chip, pass the `-DDEVICE` flag when invoking cmake. For example, to use the STM32F429Discovery, pass
+`-DDEVICE=stm32f429zit6`

--- a/src/platforms/stm32/cmake/arm-toolchain.cmake
+++ b/src/platforms/stm32/cmake/arm-toolchain.cmake
@@ -1,0 +1,13 @@
+
+# Use ARM-GCC for crosscompiling
+set(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_SYSTEM_PROCESSOR arm)
+
+find_program(ARM_CC arm-none-eabi-gcc)
+find_program(ARM_CXX arm-none-eabi-g++)
+find_program(ARM_OBJCOPY arm-none-eabi-objcopy)
+find_program(ARM_SIZE arm-none-eabi-size)
+
+SET(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
+set(CMAKE_C_COMPILER ${ARM_CC})
+set(CMAKE_CXX_COMPILER ${ARM_CXX})

--- a/src/platforms/stm32/cmake/compile-flags.cmake
+++ b/src/platforms/stm32/cmake/compile-flags.cmake
@@ -15,8 +15,7 @@ set(COMMON_WARN_FLAGS "${COMMON_WARN_FLAGS} -pedantic -Wall -Wextra \
 -Wswitch-enum -Wwrite-strings -Wundef -Wuninitialized -Wunreachable-code")
 
 # Flags that apply only to C OR C++
-set(C_WARN_FLAGS "${COMMON_WARN_FLAGS} -Wbad-function-cast -Wmissing-prototypes -Wnested-externs \
--Wold-style-definition -Wstrict-prototypes")
+set(C_WARN_FLAGS "${COMMON_WARN_FLAGS} -Wbad-function-cast -Wmissing-prototypes -Wnested-externs")
 set(CXX_WARN_FLAGS "${COMMON_WARN_FLAGS} -Wctor-dtor-privacy -Wnoexcept -Wold-style-cast -Woverloaded-virtual \
 -Wsign-promo -Wstrict-null-sentinel -Wuseless-cast -Wzero-as-null-pointer-constant")
 

--- a/src/platforms/stm32/cmake/compile-flags.cmake
+++ b/src/platforms/stm32/cmake/compile-flags.cmake
@@ -1,0 +1,31 @@
+# Enforce C/C++ standard level
+set(CMAKE_C_STANDARD_REQUIRED YES)
+set(CMAKE_CXX_STANDARD_REQUIRED YES)
+
+# Enforce to disable any compiler-specific extensions
+set(CMAKE_C_EXTENSIONS NO)
+set(CMAKE_CXX_EXTENSIONS NO)
+
+# Flags that apply to both C and C++
+set(COMMON_WARN_FLAGS "${COMMON_WARN_FLAGS} -pedantic -Wall -Wextra \
+-Wcast-align -Wcast-qual -Wconversion -Wdisabled-optimization -Wdouble-promotion -Wduplicated-cond \
+-Wduplicated-branches -Wfloat-equal -Wformat=2 -Winit-self -Winline -Winvalid-pch -Wlogical-op \
+-Wmissing-declarations -Wmissing-format-attribute -Wmissing-include-dirs -Wno-unused -Wnull-dereference -Wodr \
+-Wpointer-arith -Wredundant-decls -Wrestrict -Wshadow -Wsign-conversion -Wstrict-overflow=5 -Wswitch-default \
+-Wswitch-enum -Wwrite-strings -Wundef -Wuninitialized -Wunreachable-code")
+
+# Flags that apply only to C OR C++
+set(C_WARN_FLAGS "${COMMON_WARN_FLAGS} -Wbad-function-cast -Wmissing-prototypes -Wnested-externs \
+-Wold-style-definition -Wstrict-prototypes")
+set(CXX_WARN_FLAGS "${COMMON_WARN_FLAGS} -Wctor-dtor-privacy -Wnoexcept -Wold-style-cast -Woverloaded-virtual \
+-Wsign-promo -Wstrict-null-sentinel -Wuseless-cast -Wzero-as-null-pointer-constant")
+
+# Pass them back to the CMake variable
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${C_WARN_FLAGS}")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CXX_WARN_FLAGS}")
+
+if (LOG_VERBOSE)
+    message("-------------Warning Flags--------------")
+    message(STATUS "C   : ${C_WARN_FLAGS}")
+    message(STATUS "CXX : ${CXX_WARN_FLAGS}")
+endif ()

--- a/src/platforms/stm32/cmake/libopencm3.cmake
+++ b/src/platforms/stm32/cmake/libopencm3.cmake
@@ -1,0 +1,170 @@
+# Utility
+
+function(JOIN VALUES GLUE OUTPUT)
+    string(REGEX REPLACE "([^\\]|^);" "\\1${GLUE}" _TMP_STR "${VALUES}")
+    string(REGEX REPLACE "[\\](.)" "\\1" _TMP_STR "${_TMP_STR}") #fixes escaping
+    set(${OUTPUT} "${_TMP_STR}" PARENT_SCOPE)
+endfunction()
+
+# -----------
+
+# LibOpenCM3 Stuff
+if (NOT LIBOPENCM3_DIR)
+    find_file(LIBOPENCM3_DIR "libopencm3" "${CMAKE_CURRENT_SOURCE_DIR}" PATH_SUFFIXES)
+    if (LIBOPENCM3_DIR STREQUAL "LIBOPENCM3_DIR-NOTFOUND")
+        message(FATAL_ERROR "Could not locate libopencm3 directory")
+    endif ()
+endif ()
+
+add_custom_target(
+    libopencm3 make
+    WORKING_DIRECTORY ${LIBOPENCM3_DIR}
+)
+link_directories(${LIBOPENCM3_DIR}/lib)
+include_directories(SYSTEM ${LIBOPENCM3_DIR}/include)
+
+# Generate linker information for device, based on libopencm3/mk/genlink-config.mk
+if (NOT DEVICE)
+    message(FATAL_ERROR "No DEVICE specified for linker script generator")
+endif ()
+
+find_program(PYTHON python)
+if (NOT PYTHON)
+    message(FATAL_ERROR "python is required to generate the linker script, please install it.")
+endif ()
+
+set(GENLINK_SCRIPT "${LIBOPENCM3_DIR}/scripts/genlink.py")
+set(DEVICES_DATA "${LIBOPENCM3_DIR}/ld/devices.data")
+execute_process(
+    COMMAND "${PYTHON}" "${GENLINK_SCRIPT}" "${DEVICES_DATA}" "${DEVICE}" "FAMILY"
+    OUTPUT_VARIABLE GENLINK_FAMILY
+)
+execute_process(
+    COMMAND "${PYTHON}" "${GENLINK_SCRIPT}" "${DEVICES_DATA}" "${DEVICE}" "USBFAMILY"
+    OUTPUT_VARIABLE GENLINK_SUBFAMILY
+)
+execute_process(
+    COMMAND "${PYTHON}" "${GENLINK_SCRIPT}" "${DEVICES_DATA}" "${DEVICE}" "CPU"
+    OUTPUT_VARIABLE GENLINK_CPU
+)
+execute_process(
+    COMMAND "${PYTHON}" "${GENLINK_SCRIPT}" "${DEVICES_DATA}" "${DEVICE}" "FPU"
+    OUTPUT_VARIABLE GENLINK_FPU
+)
+execute_process(
+    COMMAND "${PYTHON}" "${GENLINK_SCRIPT}" "${DEVICES_DATA}" "${DEVICE}" "CPPFLAGS"
+    OUTPUT_VARIABLE GENLINK_CPPFLAGS
+)
+execute_process(
+    COMMAND "${PYTHON}" "${GENLINK_SCRIPT}" "${DEVICES_DATA}" "${DEVICE}" "DEFS"
+    OUTPUT_VARIABLE GENLINK_DEFS
+)
+message("----------------------------------------")
+message(STATUS "Device      : ${DEVICE}")
+message("-----------Device Linker Info-----------")
+message(STATUS "Family      : ${GENLINK_FAMILY}")
+message(STATUS "Sub-family  : ${GENLINK_SUBFAMILY}")
+message(STATUS "CPU         : ${GENLINK_CPU}")
+message(STATUS "FPU         : ${GENLINK_FPU}")
+message(STATUS "CPP Flags   : ${GENLINK_CPPFLAGS}")
+message(STATUS "Linker Defs : ${GENLINK_DEFS}")
+
+# Generate flags
+set(ARCH_FLAGS "-mcpu=${GENLINK_CPU}")
+
+# Check CPU
+set(CORTEX_CPU cortex-m0 cortex-m0plus cortex-m3 cortex-m4 cortex-m7)
+list(FILTER CORTEX_CPU INCLUDE REGEX ${GENLINK_CPU})
+if (GENLINK_CPU STREQUAL CORTEX_CPU)
+    list(APPEND ARCH_FLAGS "-mthumb -mthumb-interwork")
+endif ()
+
+# Check FPU
+if (GENLINK_FPU STREQUAL "soft")
+    list(APPEND ARCH_FLAGS "-msoft-float")
+elseif (GENLINK_FPU STREQUAL "hard-fpv4-sp-d16")
+    list(APPEND ARCH_FLAGS "-mfloat-abi=hard -mfpu=fpv4-sp-d16")
+elseif (GENLINK_FPU STREQUAL "hard-fpv5-sp-d16")
+    list(APPEND ARCH_FLAGS "-mfloat-abi=hard -mfpu=fpv5-sp-d16")
+else ()
+    message(WARNING "No match for the FPU flags")
+endif ()
+
+# Check family
+if (NOT GENLINK_FAMILY)
+    message(WARNING "${DEVICE} not found in ${DEVICES_DATA}")
+endif ()
+
+# Linker stuff
+set(LINKER_SCRIPT "generated.${DEVICE}.ld")
+
+if (EXISTS "${LIBOPENCM3_DIR}/lib/libopencm3_${GENLINK_FAMILY}.a")
+    set(LDLIBS "${LDLIBS} -lopencm3_${GENLINK_FAMILY}")
+else ()
+    if (EXISTS "${LIBOPENCM3_DIR}/lib/libopencm3_${GENLINK_SUBFAMILY}.a")
+        set(LDLIBS "${LDLIBS} -lopencm3_${GENLINK_SUBFAMILY}")
+    else ()
+        message(WARNING "${LIBOPENCM3_DIR}/lib/libopencm3_${GENLINK_FAMILY}. A library variant for the selected device does not exist.")
+    endif ()
+endif ()
+
+# ARCH_FLAGS and GENLINK_DEFS has to be passed as a list here
+string(REPLACE " " ";" GENLINK_DEFS ${GENLINK_DEFS})
+# Get rid of any spaces and turn the thing into a list
+JOIN("${ARCH_FLAGS}" " " ARCH_FLAGS)
+string(REPLACE " " ";" ARCH_FLAGS ${ARCH_FLAGS})
+# ------------------
+execute_process(
+    COMMAND ${ARM_CXX} ${ARCH_FLAGS} ${GENLINK_DEFS} "-P" "-E" "${LIBOPENCM3_DIR}/ld/linker.ld.S"
+    OUTPUT_FILE "${CMAKE_CURRENT_SOURCE_DIR}/${LINKER_SCRIPT}"
+)
+message("-----------Target Specific Info---------")
+message(STATUS "Generated Linker File   : .../${LINKER_SCRIPT}")
+
+# ARCH_FLAGS has to be passed as a string here
+JOIN("${ARCH_FLAGS}" " " ARCH_FLAGS)
+# Set linker flags
+set(LINKER_FLAGS "${LINKER_FLAGS} -specs=nosys.specs -specs=nano.specs -nostartfiles ${LDLIBS} -T${CMAKE_CURRENT_SOURCE_DIR}/${LINKER_SCRIPT} ${ARCH_FLAGS}")
+message(STATUS "Linker Flags            : ${LINKER_FLAGS}")
+
+# Compiler flags
+set(TARGET_SPECIFIC_FLAGS "${GENLINK_CPPFLAGS} ${ARCH_FLAGS}")
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${TARGET_SPECIFIC_FLAGS}")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TARGET_SPECIFIC_FLAGS}")
+
+message(STATUS "Target Specific Flags   : ${TARGET_SPECIFIC_FLAGS}")
+
+# Replace `add_executable` with custom macro with same name that adds libopencm3 as a linking target
+macro(add_executable _name)
+    # invoke built-in add_executable
+    if ("${ARGV1}" STREQUAL "")
+        message(FATAL_ERROR "No source files added to executable")
+    endif ()
+    _add_executable(${ARGV})
+
+    if (TARGET ${_name} AND NOT ${_name} MATCHES ".*Doxygen.*")
+        # Set target properties
+        set_target_properties(${ARGV0} PROPERTIES LINK_FLAGS ${LINKER_FLAGS})
+        set_target_properties(${PROJECT_NAME}.elf PROPERTIES OUTPUT_NAME "${PROJECT_NAME}.elf")
+        target_link_libraries(${_name} "opencm3_${GENLINK_FAMILY}")
+
+        # Set output file locations
+        string(REGEX MATCH "^(.*)\\.[^.]*$" dummy "${ARGV0}")
+        set(bin ${CMAKE_MATCH_1}.bin)
+        set(elf ${ARGV0})
+        set(bin_out ${CMAKE_MATCH_1}.bin)
+        get_target_property(elf_in ${PROJECT_NAME}.elf OUTPUT_NAME)
+
+        # Add target
+        add_custom_target(
+            ${bin} ALL
+            COMMAND ${ARM_OBJCOPY} -Obinary ${elf_in} ${bin_out}
+            WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}
+            DEPENDS ${elf}
+        )
+
+        message("------------Output Locations------------")
+        message(STATUS "BIN: ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${bin_out}")
+        message(STATUS "ELF: ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${elf_in}")
+    endif ()
+endmacro()

--- a/src/platforms/stm32/src/gpiodriver.c
+++ b/src/platforms/stm32/src/gpiodriver.c
@@ -1,0 +1,176 @@
+/***************************************************************************
+ *   Copyright 2018 by Riccardo Binetti <rbino@gmx.com>                    *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU Lesser General Public License as        *
+ *   published by the Free Software Foundation; either version 2 of the    *
+ *   License, or (at your option) any later version.                       *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
+ ***************************************************************************/
+
+#include "gpiodriver.h"
+
+#include <libopencm3/stm32/gpio.h>
+#include <libopencm3/stm32/rcc.h>
+
+#include <context.h>
+#include <mailbox.h>
+#include <term.h>
+
+#include <trace.h>
+
+static void consume_gpio_mailbox(Context *ctx);
+static uint32_t port_atom_to_gpio_port(Context *ctx, term port_atom);
+static uint16_t gpio_port_to_rcc_port(uint32_t gpio_port);
+static char gpio_port_to_name(uint32_t gpio_port);
+
+static const char *const ok_a = "\x2" "ok";
+static const char *const error_a = "\x5" "error";
+static const char *const set_level_a = "\x9" "set_level";
+static const char *const input_a = "\x5" "input";
+static const char *const output_a = "\x6" "output";
+static const char *const set_direction_a ="\xD" "set_direction";
+static const char *const a_a = "\x01" "a";
+static const char *const b_a = "\x01" "b";
+static const char *const c_a = "\x01" "c";
+static const char *const d_a = "\x01" "d";
+static const char *const e_a = "\x01" "e";
+static const char *const f_a = "\x01" "f";
+
+
+void gpiodriver_init(Context *ctx)
+{
+    ctx->native_handler = consume_gpio_mailbox;
+    ctx->platform_data = NULL;
+}
+
+static void consume_gpio_mailbox(Context *ctx)
+{
+    term ret;
+
+    Message *message = mailbox_dequeue(ctx);
+    term msg = message->message;
+    term pid = term_get_tuple_element(msg, 0);
+    term cmd = term_get_tuple_element(msg, 1);
+
+    int local_process_id = term_to_local_process_id(pid);
+    Context *target = globalcontext_get_process(ctx->global, local_process_id);
+
+    if (cmd == context_make_atom(ctx, set_level_a)) {
+        term gpio_tuple = term_get_tuple_element(msg, 2);
+        term gpio_port_atom = term_get_tuple_element(gpio_tuple, 0);
+        uint32_t gpio_port = port_atom_to_gpio_port(ctx, gpio_port_atom);
+        int32_t gpio_pin_num = term_to_int32(term_get_tuple_element(gpio_tuple, 1));
+        int32_t level = term_to_int32(term_get_tuple_element(msg, 3));
+
+        if (level != 0) {
+            gpio_set(gpio_port, 1 << gpio_pin_num);
+        } else {
+            gpio_clear(gpio_port, 1 << gpio_pin_num);
+        }
+        TRACE("gpio: set_level: %c%i %i\n", gpio_port_to_name(gpio_port), gpio_pin_num, level != 0);
+        ret = context_make_atom(ctx, ok_a);
+
+    } else if (cmd == context_make_atom(ctx, set_direction_a)) {
+        term gpio_tuple = term_get_tuple_element(msg, 2);
+        term gpio_port_atom = term_get_tuple_element(gpio_tuple, 0);
+        uint32_t gpio_port = port_atom_to_gpio_port(ctx, gpio_port_atom);
+        int32_t gpio_pin_num = term_to_int32(term_get_tuple_element(gpio_tuple, 1));
+        term direction = term_get_tuple_element(msg, 3);
+
+        uint16_t rcc_port = gpio_port_to_rcc_port(gpio_port);
+        // Set direction implicitly enables the port of the GPIO
+        rcc_periph_clock_enable(rcc_port);
+
+        if (direction == context_make_atom(ctx, input_a)) {
+            gpio_mode_setup(gpio_port, GPIO_MODE_INPUT, GPIO_PUPD_NONE, 1 << gpio_pin_num);
+            TRACE("gpio: set_direction: %c%i INPUT\n", gpio_port_to_name(gpio_port), gpio_pin_num);
+            ret = context_make_atom(ctx, ok_a);
+
+        } else if (direction == context_make_atom(ctx, output_a)) {
+            gpio_mode_setup(gpio_port, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, 1 << gpio_pin_num);
+            TRACE("gpio: set_direction: %c%i OUTPUT\n", gpio_port_to_name(gpio_port), gpio_pin_num);
+            ret = context_make_atom(ctx, ok_a);
+
+        } else {
+            TRACE("gpio: unrecognized direction\n");
+            ret = context_make_atom(ctx, error_a);
+        }
+
+    } else {
+        TRACE("gpio: unrecognized command\n");
+        ret = context_make_atom(ctx, error_a);
+    }
+
+    free(message);
+
+    mailbox_send(target, ret);
+}
+
+static uint32_t port_atom_to_gpio_port(Context *ctx, term port_atom)
+{
+    if (port_atom == context_make_atom(ctx, a_a)) {
+        return GPIOA;
+    } else if (port_atom == context_make_atom(ctx, b_a)) {
+        return GPIOB;
+    } else if (port_atom == context_make_atom(ctx, c_a)) {
+        return GPIOC;
+    } else if (port_atom == context_make_atom(ctx, d_a)) {
+        return GPIOD;
+    } else if (port_atom == context_make_atom(ctx, e_a)) {
+        return GPIOE;
+    } else if (port_atom == context_make_atom(ctx, f_a)) {
+        return GPIOF;
+    } else {
+        return 0;
+    }
+}
+
+static uint16_t gpio_port_to_rcc_port(uint32_t gpio_port)
+{
+    switch (gpio_port) {
+        case GPIOA:
+            return RCC_GPIOA;
+        case GPIOB:
+            return RCC_GPIOB;
+        case GPIOC:
+            return RCC_GPIOC;
+        case GPIOD:
+            return RCC_GPIOD;
+        case GPIOE:
+            return RCC_GPIOE;
+        case GPIOF:
+            return RCC_GPIOF;
+        default:
+            return 0;
+    }
+}
+
+static char gpio_port_to_name(uint32_t gpio_port)
+{
+    switch (gpio_port) {
+        case GPIOA:
+            return 'A';
+        case GPIOB:
+            return 'B';
+        case GPIOC:
+            return 'C';
+        case GPIOD:
+            return 'D';
+        case GPIOE:
+            return 'E';
+        case GPIOF:
+            return 'F';
+        default:
+            return 0;
+    }
+}

--- a/src/platforms/stm32/src/gpiodriver.h
+++ b/src/platforms/stm32/src/gpiodriver.h
@@ -1,0 +1,27 @@
+/***************************************************************************
+ *   Copyright 2018 by Riccardo Binetti <rbino@gmx.com>                    *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU Lesser General Public License as        *
+ *   published by the Free Software Foundation; either version 2 of the    *
+ *   License, or (at your option) any later version.                       *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
+ ***************************************************************************/
+
+#ifndef _GPIODRIVER_H_
+#define _GPIODRIVER_H_
+
+#include <context.h>
+
+void gpiodriver_init(Context *ctx);
+
+#endif

--- a/src/platforms/stm32/src/main.c
+++ b/src/platforms/stm32/src/main.c
@@ -1,0 +1,23 @@
+/***************************************************************************
+ *   Copyright 2018 by Riccardo Binetti <rbino@gmx.com>                    *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU Lesser General Public License as        *
+ *   published by the Free Software Foundation; either version 2 of the    *
+ *   License, or (at your option) any later version.                       *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
+ ***************************************************************************/
+
+int main()
+{
+    return 42;
+}

--- a/src/platforms/stm32/src/main.c
+++ b/src/platforms/stm32/src/main.c
@@ -17,7 +17,74 @@
  *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
  ***************************************************************************/
 
+#include <errno.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <libopencm3/stm32/gpio.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/stm32/usart.h>
+
+#define USART_CONSOLE USART2
+
+int _write(int file, char *ptr, int len);
+
+static void clock_setup()
+{
+    // Use external clock, set divider for 168 MHz clock frequency
+    rcc_clock_setup_hse_3v3(&rcc_hse_8mhz_3v3[RCC_CLOCK_3V3_168MHZ]);
+
+    // Enable clock for USART2 GPIO
+    rcc_periph_clock_enable(RCC_GPIOA);
+
+    // Enable clock for USART2
+    rcc_periph_clock_enable(RCC_USART2);
+}
+
+static void usart_setup()
+{
+    // Setup GPIO pins for USART2 transmit
+    gpio_mode_setup(GPIOA, GPIO_MODE_AF, GPIO_PUPD_NONE, GPIO2);
+
+    // Setup USART2 TX pin as alternate function
+    gpio_set_af(GPIOA, GPIO_AF7, GPIO2);
+
+    usart_set_baudrate(USART_CONSOLE, 115200);
+    usart_set_databits(USART_CONSOLE, 8);
+    usart_set_stopbits(USART_CONSOLE, USART_STOPBITS_1);
+    usart_set_mode(USART_CONSOLE, USART_MODE_TX);
+    usart_set_parity(USART_CONSOLE, USART_PARITY_NONE);
+    usart_set_flow_control(USART_CONSOLE, USART_FLOWCONTROL_NONE);
+
+    // Finally enable the USART
+    usart_enable(USART_CONSOLE);
+}
+
+// Use USART_CONSOLE as a console.
+// This is a syscall for newlib
+int _write(int file, char *ptr, int len)
+{
+    int i;
+
+    if (file == STDOUT_FILENO || file == STDERR_FILENO) {
+        for (i = 0; i < len; i++) {
+            if (ptr[i] == '\n') {
+                usart_send_blocking(USART_CONSOLE, '\r');
+            }
+            usart_send_blocking(USART_CONSOLE, ptr[i]);
+        }
+        return i;
+    }
+    errno = EIO;
+    return -1;
+}
+
 int main()
 {
-    return 42;
+    clock_setup();
+    usart_setup();
+    printf("\nBooting AtomVM\n");
+
+    while (1);
+
+    return 0;
 }

--- a/src/platforms/stm32/src/main.c
+++ b/src/platforms/stm32/src/main.c
@@ -26,7 +26,14 @@
 #include <libopencm3/stm32/rcc.h>
 #include <libopencm3/stm32/usart.h>
 
+#include <avmpack.h>
+#include <context.h>
+#include <globalcontext.h>
+#include <module.h>
+
 #define USART_CONSOLE USART2
+#define AVM_ADDRESS (0x8080000)
+#define AVM_FLASH_MAX_SIZE (0x80000)
 #define CLOCK_FREQUENCY (168000000)
 
 int _write(int file, char *ptr, int len);
@@ -97,7 +104,34 @@ int main()
     clock_setup();
     systick_setup();
     usart_setup();
-    printf("\nBooting AtomVM\n");
+
+    const void *flashed_avm = (void *) AVM_ADDRESS;
+    uint32_t size = AVM_FLASH_MAX_SIZE;
+
+    uint32_t startup_beam_size;
+    const void *startup_beam;
+    const char *startup_module_name;
+
+    printf("Booting file mapped at: %p, size: %li\n", flashed_avm, size);
+
+    GlobalContext *glb = globalcontext_new();
+
+    if (!avmpack_is_valid(flashed_avm, size) || !avmpack_find_section_by_flag(flashed_avm, BEAM_START_FLAG, &startup_beam, &startup_beam_size, &startup_module_name)) {
+        fprintf(stderr, "error: invalid AVM Pack\n");
+        return 1;
+    }
+
+    glb->avmpack_data = flashed_avm;
+    glb->avmpack_platform_data = NULL;
+
+    Module *mod = module_new_from_iff_binary(glb, startup_beam, startup_beam_size);
+    globalcontext_insert_module_with_filename(glb, mod, startup_module_name);
+    Context *ctx = context_new(glb);
+
+    printf("Starting: %s...\n", startup_module_name);
+    printf("---\n");
+    context_execute_loop(ctx, mod, "start", 0);
+    printf("Return value: %lx\n", (long) term_to_int32(ctx->x[0]));
 
     while (1);
 

--- a/src/platforms/stm32/src/main.c
+++ b/src/platforms/stm32/src/main.c
@@ -20,11 +20,14 @@
 #include <errno.h>
 #include <stdio.h>
 #include <unistd.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/cm3/systick.h>
 #include <libopencm3/stm32/gpio.h>
 #include <libopencm3/stm32/rcc.h>
 #include <libopencm3/stm32/usart.h>
 
 #define USART_CONSOLE USART2
+#define CLOCK_FREQUENCY (168000000)
 
 int _write(int file, char *ptr, int len);
 
@@ -59,6 +62,17 @@ static void usart_setup()
     usart_enable(USART_CONSOLE);
 }
 
+// Set up a timer to create 1ms ticks
+// The handler is in sys.c
+static void systick_setup()
+{
+    // clock rate / 1000 to get 1ms interrupt rate
+    systick_set_reload(CLOCK_FREQUENCY / 1000);
+    systick_set_clocksource(STK_CSR_CLKSOURCE_AHB);
+    systick_counter_enable();
+    systick_interrupt_enable();
+}
+
 // Use USART_CONSOLE as a console.
 // This is a syscall for newlib
 int _write(int file, char *ptr, int len)
@@ -81,6 +95,7 @@ int _write(int file, char *ptr, int len)
 int main()
 {
     clock_setup();
+    systick_setup();
     usart_setup();
     printf("\nBooting AtomVM\n");
 

--- a/src/platforms/stm32/src/platform.c
+++ b/src/platforms/stm32/src/platform.c
@@ -1,0 +1,26 @@
+/***************************************************************************
+ *   Copyright 2018 by Riccardo Binetti <rbino@gmx.com>                    *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU Lesser General Public License as        *
+ *   published by the Free Software Foundation; either version 2 of the    *
+ *   License, or (at your option) any later version.                       *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
+ ***************************************************************************/
+
+#include <context.h>
+
+Context *platform_open_port(GlobalContext *glb, const char *driver_name, term opts)
+{
+    // TODO: implement
+    return NULL;
+}

--- a/src/platforms/stm32/src/platform.c
+++ b/src/platforms/stm32/src/platform.c
@@ -18,9 +18,18 @@
  ***************************************************************************/
 
 #include <context.h>
+#include <gpiodriver.h>
 
 Context *platform_open_port(GlobalContext *glb, const char *driver_name, term opts)
 {
-    // TODO: implement
-    return NULL;
+    Context *new_ctx = context_new(glb);
+
+    if (!strcmp(driver_name, "gpio")) {
+        gpiodriver_init(new_ctx);
+    } else {
+        context_destroy(new_ctx);
+        return NULL;
+    }
+
+    return new_ctx;
 }

--- a/src/platforms/stm32/src/sys.c
+++ b/src/platforms/stm32/src/sys.c
@@ -1,0 +1,39 @@
+/***************************************************************************
+ *   Copyright 2018 by Riccardo Binetti <rbino@gmx.com>                    *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU Lesser General Public License as        *
+ *   published by the Free Software Foundation; either version 2 of the    *
+ *   License, or (at your option) any later version.                       *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
+ ***************************************************************************/
+
+#include <sys.h>
+
+// Monotonically increasing number of milliseconds from reset
+// Overflows every 49 days
+// TODO: use 64 bit (remember to take into account atomicity)
+static volatile uint32_t system_millis;
+
+// Called when systick fires
+void sys_tick_handler()
+{
+    system_millis++;
+}
+
+// Sleep for delay milliseconds
+static void msleep(uint32_t delay)
+{
+    // TODO: use a smarter sleep instead of busy waiting
+    uint32_t wake = system_millis + delay;
+    while (wake > system_millis);
+}

--- a/src/platforms/stm32/src/sys.c
+++ b/src/platforms/stm32/src/sys.c
@@ -19,6 +19,9 @@
 
 #include <sys.h>
 
+#include <avmpack.h>
+#include <scheduler.h>
+
 // Monotonically increasing number of milliseconds from reset
 // Overflows every 49 days
 // TODO: use 64 bit (remember to take into account atomicity)
@@ -36,4 +39,97 @@ static void msleep(uint32_t delay)
     // TODO: use a smarter sleep instead of busy waiting
     uint32_t wake = system_millis + delay;
     while (wake > system_millis);
+}
+
+static inline void sys_clock_gettime(struct timespec *t)
+{
+    t->tv_sec = system_millis / 1000;
+    t->tv_nsec = (system_millis % 1000) * 1000000;
+}
+
+static int32_t timespec_diff_to_ms(struct timespec *timespec1, struct timespec *timespec2)
+{
+    return (timespec1->tv_sec - timespec2->tv_sec) * 1000 + (timespec1->tv_nsec - timespec2->tv_nsec) / 1000000;
+}
+
+void sys_waitevents(struct ListHead *listeners_list)
+{
+    struct timespec now;
+    sys_clock_gettime(&now);
+
+    EventListener *listeners = GET_LIST_ENTRY(listeners_list, EventListener, listeners_list_head);
+
+    int min_timeout = INT_MAX;
+    int count = 0;
+
+    //first: find maximum allowed sleep time, and count file descriptor listeners
+    EventListener *listener = listeners;
+    do {
+        if (listener->expires) {
+            int wait_ms = timespec_diff_to_ms(&listener->expiral_timestamp, &now);
+            if (wait_ms <= 0) {
+                min_timeout = 0;
+            } else if (min_timeout > wait_ms) {
+                min_timeout = wait_ms;
+            }
+        }
+        if (listener->fd >= 0) {
+            count++;
+        }
+
+        listener = GET_LIST_ENTRY(listener->listeners_list_head.next, EventListener, listeners_list_head);
+    } while (listener != listeners);
+
+    msleep(min_timeout);
+
+    //second: execute handlers for expiered timers
+    if (min_timeout != INT_MAX) {
+        listener = listeners;
+        sys_clock_gettime(&now);
+        do {
+            EventListener *next_listener = GET_LIST_ENTRY(listener->listeners_list_head.next, EventListener, listeners_list_head);
+            if (listener->expires) {
+                int wait_ms = timespec_diff_to_ms(&listener->expiral_timestamp, &now);
+                if (wait_ms <= 0) {
+                    //it is completely safe to free a listener in the callback, we are going to not use it after this call
+                    listener->handler(listener);
+                }
+            }
+
+            listener = next_listener;
+        } while (listener != listeners);
+    }
+}
+
+void sys_set_timestamp_from_relative_to_abs(struct timespec *t, int32_t millis)
+{
+    sys_clock_gettime(t);
+    t->tv_sec += millis / 1000;
+    t->tv_nsec += (millis % 1000) * 1000000;
+}
+
+void sys_time(struct timespec *t)
+{
+    sys_clock_gettime(t);
+}
+
+Module *sys_load_module(GlobalContext *global, const char *module_name)
+{
+    const void *beam_module = NULL;
+    uint32_t beam_module_size = 0;
+
+    if (!(global->avmpack_data && avmpack_find_section_by_name(global->avmpack_data, module_name, &beam_module, &beam_module_size))) {
+        fprintf(stderr, "Failed to open module: %s\n", module_name);
+        return NULL;
+    }
+
+    Module *new_module = module_new_from_iff_binary(global, beam_module, beam_module_size);
+    new_module->module_platform_data = NULL;
+
+    return new_module;
+}
+
+void sys_platform_periodic_tasks()
+{
+    // TODO: implement
 }

